File: R/resnet-conv-block.R
relu <- function(x) {
  pmax(0, x)
}

ConvBlock <- setRefClass(
  "ConvBlock",
  fields = list(
    in_channels = "numeric",
    out_channels = "numeric",
    W1 = "matrix",
    W2 = "matrix",
    Ws = "matrix"
  ),
  methods = list(
    initialize = function(in_channels, out_channels) {
      in_channels <<- in_channels
      out_channels <<- out_channels
      W1 <<- matrix(rnorm(in_channels * out_channels, sd = 0.01), nrow = in_channels, ncol = out_channels)
      W2 <<- matrix(rnorm(out_channels * out_channels, sd = 0.01), nrow = out_channels, ncol = out_channels)
      Ws <<- matrix(rnorm(in_channels * out_channels, sd = 0.01), nrow = in_channels, ncol = out_channels)
    },
    forward = function(x) {
      main <- relu(x %*% W1)
      main <- main %*% W2
      shortcut <- x %*% Ws
      relu(main + shortcut)
    }
  )
)

File: Julia/resnet-conv-block.jl
relu(x) = max.(0, x)

mutable struct ConvBlock
    in_channels::Int
    out_channels::Int
    W1
    W2
    Ws
end

function ConvBlock(in_channels::Int, out_channels::Int)
    W1 = randn(in_channels, out_channels) .* 0.01
    W2 = randn(out_channels, out_channels) .* 0.01
    Ws = randn(in_channels, out_channels) .* 0.01
    ConvBlock(in_channels, out_channels, W1, W2, Ws)
end

function forward(block::ConvBlock, x)
    main = relu(x * block.W1)
    main = main * block.W2
    shortcut = x * block.Ws
    relu(main .+ shortcut)
end
